pkg/receive/writer_test.go

```diff
@@ -174,3 +175,80 @@ func TestWriter(t *testing.T) {
 		})
 	}
 }
+
+func BenchmarkWriterTimeSeriesWithSingleLabel_10(b *testing.B)   { benchmarkWriter(b, 1, 10) }
+func BenchmarkWriterTimeSeriesWithSingleLabel_100(b *testing.B)  { benchmarkWriter(b, 1, 100) }
+func BenchmarkWriterTimeSeriesWithSingleLabel_1000(b *testing.B) { benchmarkWriter(b, 1, 1000) }
+
+func BenchmarkWriterTimeSeriesWith10Labels_10(b *testing.B)   { benchmarkWriter(b, 10, 10) }
+func BenchmarkWriterTimeSeriesWith10Labels_100(b *testing.B)  { benchmarkWriter(b, 10, 100) }
+func BenchmarkWriterTimeSeriesWith10Labels_1000(b *testing.B) { benchmarkWriter(b, 10, 1000) }
+
+func benchmarkWriter(b *testing.B, labelsNum int, seriesNum int) {
+	dir, err := ioutil.TempDir("", "bench-writer")
+	testutil.Ok(b, err)
+	defer func() { testutil.Ok(b, os.RemoveAll(dir)) }()
+
+	logger := log.NewNopLogger()
+
+	m := NewMultiTSDB(dir, logger, prometheus.NewRegistry(), &tsdb.Options{
+		MinBlockDuration:      (2 * time.Hour).Milliseconds(),
+		MaxBlockDuration:      (2 * time.Hour).Milliseconds(),
+		RetentionDuration:     (6 * time.Hour).Milliseconds(),
+		NoLockfile:            true,
+		MaxExemplars:          0,
+		EnableExemplarStorage: true,
+	},
+		labels.FromStrings("replica", "01"),
+		"tenant_id",
+		nil,
+		false,
+		metadata.NoneFunc,
+	)
+	defer func() { testutil.Ok(b, m.Close()) }()
+
+	testutil.Ok(b, m.Flush())
+	testutil.Ok(b, m.Open())
+
+	app, err := m.TenantAppendable("foo")
+	testutil.Ok(b, err)
+
+	w := NewWriter(logger, m)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	testutil.Ok(b, runutil.Retry(1*time.Second, ctx.Done(), func() error {
+		_, err = app.Appender(context.Background())
+		return err
+	}))
+
+	timeSeries := generateLabelsAndSeries(labelsNum, seriesNum)
+
+	wreq := &prompb.WriteRequest{
+		Timeseries: timeSeries,
+	}
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		_ = w.Write(ctx, "foo", wreq)
+	}
+}
+
+func generateLabelsAndSeries(numLabels int, numSeries int) []prompb.TimeSeries {
+	// Generate some labels first.
+	l := make([]labelpb.ZLabel, 0, numLabels)
+	l = append(l, labelpb.ZLabel{Name: "__name__", Value: "test"})
+	for i := 0; i < numLabels; i++ {
+		l = append(l, labelpb.ZLabel{Name: fmt.Sprintf("label_%q", rune('a'-1+i)), Value: fmt.Sprintf("%d", i)})
+	}
+
+	ts := make([]prompb.TimeSeries, 0, numSeries)
+	for j := 0; j < numSeries; j++ {
+		ts = append(ts, prompb.TimeSeries{Labels: l, Samples: []prompb.Sample{{Value: 1, Timestamp: 10}}})
```

Comment:
isn't the same timestamp a problem? I thought we should be getting conflict in append.

Reply:
Unfortunately either we have to mock the check for this in TSDB (hard) or create / update ts on every write.

Reply:
So I'm not 100% sure, but it looks it has to do something how Prometheus [detects duplicates](https://github.com/prometheus/prometheus/blob/main/tsdb/head_append.go#L342). Even with an error check, the current version of benchmark does not error out. If I manipulated the timestamps to be out of order, I was seeing the conflict errors in benchmark. I'll adjust the samples to be for different timestamps, but technically it seems the behavior of the benchmark is correct (error free).

Reply:
Great! I would say then let's say comment clearly somewhere that we DON'T include potential overhead (CPU, mem) of TSDB append as we don't append anything.

Reply:
Will do :+1: